### PR TITLE
Add cohere model provider

### PR DESCRIPTION
**Pull request by arkadyark-cohere**
Hey @gkamradt! I'm Arkady from Cohere - we'd love to get our models into here so our results can be more easily reproduced and verified. This PR adds support for Cohere models for easier reproducibility of our reported results.

Ran with `needlehaystack.run_test --provider cohere --model_name command-r --evaluator_model_name "gpt-4" --document_depth_percents "[0,11,22,33,44,56,67,78,89,100]" --context_lengths "[1000,10000,19000,28000,37000,46000,55000,64000,73000,82000,91000,100000,109000,118000,128000]"`

image

Some results slightly differing from our reported, due to:
- Some stochasticity, we generated the context 8 times with random seeds and averaged
- Small differences in the public cohere SDK v5, vs our internal evaluation
- Most of the 7/10s come from the GPT-4 evaluator being harsh on a response like `Eat a sandwich and sit in Dolores Park on a sunny day.`

But overall, nothing major.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -6,6 +6,7 @@ anyio==3.7.1
 attrs==23.1.0
 certifi==2023.11.17
 charset-normalizer==3.3.2
+cohere>=5.0.0
```
Comment (daulet) [inline requirements.txt:9]: let's use `>=5.1.2`

File: needlehaystack/providers/cohere.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        self.model_kwargs = model_kwargs
+        self.api_key = api_key
+
+        self.client = AsyncClient(api_key=self.api_key)
+
+    async def evaluate_model(self, prompt: str) -> str:
+        response = await self.client.chat(message=prompt[-1]["message"], chat_history=prompt[:-1], model=self.model_name, **self.model_kwargs)
+        return response.text
+
+    def generate_prompt(self, context: str, retrieval_question: str) -> str | list[dict[str, str]]:
```
Comment (mkozakov) [inline needlehaystack/providers/cohere.py:37]: would it be cleaner if this returned a touple of chat_history and message? that way you don't have to do the `-1` index acrobatics

File: requirements.txt (hunk 1 above)
Comment (arkadyark-cohere) [reply requirements.txt]: Done, ty

File: needlehaystack/providers/cohere.py (hunk 2 above)
Comment (arkadyark-cohere) [reply needlehaystack/providers/cohere.py]: Good call, ty

**Comment by arkadyark-cohere** [comment]
hey @gkamradt @pavelkraleu - is there anybody available to potentially take a look at getting this in? We'd love to be able to externally reproduce our numbers :)

**Comment by gkamradt** [comment]
Just ran it, looks great - thank you!

**Comment by kedarchandrayan** [comment]
Looks good to me. Merging it.

**Review by Reality2byte — APPROVED** [review]

**Pull request merged**